### ISSUE TITLE
Dropping Window before Swapchain can lock up program

### ISSUE THREAD

**Issue by branan** [vulkano-rs/vulkano#844]
Currently, the `Swapchain` and `Window` objects are totally unrelated. This means that it's possible to drop a `Window` while a `Swapchain` derived from it is still in use. At least under the Linux Intel driver (17.0.7), this will cause the driver to lock up in `vkDestroySwapchainKHR`, waiting on an event for a window that no longer exists.

When all of your Vulkan objects are function-local variables, this drop order is implicitly going to be "right" most of the time (the opposite order you initialized them in). When storing them in structs, though, it's easy to get wrong. The following struct will lock up when it is dropped:

```
struct VkObjects {
    previous_frame_end: Box<GpuFuture>,
    framebuffers: Option<Vec<Arc<FramebufferAbstract + Send + Sync>>>,
    render_pass: Arc<RenderPassAbstract + Send + Sync>,
    images: Vec<Arc<SwapchainImage>>,
    window: Window,
    swapchain: Arc<Swapchain>,
    queue: Arc<Queue>,
    device: Arc<Device>,
    event_loop: EventsLoop,
}
```

I'm unsure if this is a bug in Vulkano or the driver (or just in my code). I couldn't find anything in the Vulkan spec that says a native window must outlive its associated `VkSurface`, but I'm also not surprised that the driver makes this assumption.

This can *probably* be solved with a new `WindowAbstract` trait that can be `Arc`'d by the `Surface`, and another layer of indirection around window handling. I'm willing to do that work if you're interested in that approach.

**Comment by GabrielMajeri** [comment]
The [Vulkano docs](https://docs.rs/vulkano/0.7.1/vulkano/swapchain/index.html#creating-a-surface-from-a-window) seem to be pretty clear:

> **Note that the Surface object is unsafe.** It is your responsibility to keep the window alive for at least as long as the surface exists.

Since the swapchain is pretty much tied to the surface, it is also, in extensio, **your** responsability.

As to whether the Intel driver should not deadlock, well the spec says something like this:

> In general, objects **can** be destroyed or freed in any order

Although it is talking about Vulkan objects, and a window is technically not a Vulkan object (but the surface is).

It is still @tomaka's call whether a higher-level window trait should (eventually) exist.

**Comment by tomaka** [comment]
Thanks for opening an issue.
[I noticed this problem some time ago](https://github.com/vulkano-rs/vulkano/blob/626863d7c0ed384f88fa7a28021052f04c4cf659/vulkano/src/swapchain/swapchain.rs#L222), unfortunately fixing it requires breaking changes, and I forgot to do it before releasing 0.7.

**Comment by tomaka** [comment]
I'm also a bit unsure exactly *how* to fix it.

**Comment by branan** [comment]
@GabrielMajeri Totally agreed - and there are lots of ways to deadlock safe rust programs - but I spent enough time pulling my hair out I figured it was worth raising a ticket, even if "user error" is the root cause. I'm actually on Mir (ubuntu 17.04).

@tomaka My idea goes something like this (I haven't written any of the real code yet, so it's possible I'm missing something):

```rust
// in vulkano::swapchain
trait WindowAbstract;

struct Surface {
    // fields omitted
    _window: Arc<WindowAbstract>,
}

// in vulkano_win
struct WinitWindow (winit::Window);

impl WindowAbstract for WinitWindow {}

struct Window {
    window: Arc<WinitWindow>,
    surface: Arc<Surface>
}
```

The surface can now ref its window through an opaque trait. We still need the `vulkano_win::Window` type to exist so that we have a way to access the `winit::Window` through its real type, instead of the opaque Arc in the surface. It'd be nice to just have the user own the `Arc<Surface>` directly here and drop the Window struct entirely, but I wouldn't want to add a full window interface trait inside `vulkano::swapchain`

The big problem with this is that it's clunky for users. Maybe it'd be possible to `impl deref<winit::Window> for WinitWindow` to avoid some of that?

**Comment by branan** [comment]
I got bored over lunch so took a pass at this. The two referenced commits above are different approaches to integrating what I described. Both seem to work correctly to resolve this issues. The first changes public API, and would require a 0.8.0 release. The second keeps the API stable, but uses a bit of `unsafe` to do so. I'd prefer the first, as I don't think there's any reason to push this in a Z release.

I've got to get back to my real job, but when I have a chance to clean it up, I'll open a PR.
